File: runtime/attachments.py
"""Per-conversation file attachments: validation, storage, text extraction
(with OCR), indexing for on-demand retrieval, and expiry.

Scope decided with the product: jpg/jpeg/pdf/txt/md only, one file at a time,
5 MB maximum. This web runtime is the test and benchmarking surface; the
feature ships to users in the mobile app afterwards, so everything here is
kept device-shaped — plain files on disk, SQLite, a subprocess OCR engine —
rather than leaning on anything server-only.

Pipeline for one upload:
  validate  -> extension AND magic bytes must agree; size cap; no NUL bytes
               in text files (a renamed binary is refused, not decoded).
  extract   -> txt/md: decode.  pdf: pypdf text per page, and any page with
               almost no text layer is rasterised (PyMuPDF, 200 dpi) and OCR'd.
               jpg: Tesseract OCR after light preprocessing.
  index     -> chunk (core.chunking, section-aware) + embed (bge-small) into a
               kb.sqlite beside the file, so retrieval at question time is the
               same code path as the agent's own knowledge base.
  guard     -> runtime/guard.py scans the extracted text; the verdict is stored
               in meta.json so every later question against this file is
               refused without re-scanning.

"On-demand RAG": a small document (<= FULLTEXT_CHARS) is injected whole at
question time — retrieval over three chunks only loses information. Larger
documents go through the lexical gate and vector search exactly like the KB.

Retention: attachments are conversation-scoped user data, not knowledge. They
are deleted on request and swept after RETENTION_HOURS regardless.
"""
from __future__ import annotations

import hashlib
import io
import json
import os
import re
import shutil
import subprocess
import time
import uuid
from pathlib import Path

from core import chunking, embeddings, gating, kbstore

APP_DIR = Path(__file__).resolve().parent
ATTACH_DIR = APP_DIR / "device_storage" / "attachments"
ATTACH_DIR.mkdir(parents=True, exist_ok=True)

ALLOWED_EXT = {".jpg": "image", ".jpeg": "image", ".pdf": "pdf", ".txt": "text", ".md": "text"}
MAX_BYTES = 5 * 1024 * 1024
RETENTION_HOURS = 24

FULLTEXT_CHARS = 3500        # inject whole; ~3 chunks / ~900 tokens
ATTACH_TOP_K = 4
ATTACH_MIN_SCORE = 0.30      # a floor only — the user chose this file, so lean inclusive
MIN_TEXT_LAYER_CHARS = 40    # a PDF page below this is treated as scanned -> OCR
OCR_DPI = 300                # Tesseract's recommended density; 200 misread "30" as "3@"
MAX_OCR_PAGES = 40           # 5 MB of scanned pages could be 100+; cap the wall-clock
OCR_TIMEOUT_S = 60

TESSERACT = os.environ.get("TESSERACT_EXE") or r"C:\Program Files\Tesseract-OCR\tesseract.exe"


class AttachmentError(ValueError):
    def __init__(self, message: str, status: int = 400):
        super().__init__(message)
        self.status = status


# ------------------------------------------------------------------ validation

def validate(filename: str, data: bytes) -> tuple[str, str]:
    """Return (ext, kind) or raise AttachmentError. Type is decided by the
    extension AND the file's own bytes — a .txt that is really an executable,
    or a .jpg that is really a PDF, is refused rather than guessed at."""
    name = (filename or "").strip()
    if not name:
        raise AttachmentError("the file has no name")
    ext = Path(name).suffix.lower()
    if ext not in ALLOWED_EXT:
        raise AttachmentError(
            f"'{ext or name}' is not supported — attach a jpg, jpeg, pdf, txt or md file")
    if not data:
        raise AttachmentError("the file is empty")
    if len(data) > MAX_BYTES:
        raise AttachmentError(
            f"file is {len(data) / (1024 * 1024):.1f} MB; the limit is 5 MB", status=413)
    kind = ALLOWED_EXT[ext]
    if kind == "image" and not data.startswith(b"\xff\xd8\xff"):
        raise AttachmentError("this is not a JPEG image (the bytes do not match the extension)")
    if kind == "pdf" and not data[:1024].lstrip().startswith(b"%PDF-"):
        raise AttachmentError("this is not a PDF (the bytes do not match the extension)")
    if kind == "text":
        sample = data[:65536]
        if b"\x00" in sample:
            raise AttachmentError("this is not a text file (binary content)")
    return ext, kind


# ------------------------------------------------------------------ storage

def _dir(att_id: str) -> Path:
    if not re.fullmatch(r"[0-9a-f]{32}", att_id or ""):
        raise AttachmentError("unknown attachment", status=404)
    return ATTACH_DIR / att_id


def save(filename: str, data: bytes) -> dict:
    ext, kind = validate(filename, data)
    att_id = uuid.uuid4().hex
    d = _dir(att_id)
    d.mkdir(parents=True, exist_ok=False)
    (d / f"file{ext}").write_bytes(data)
    meta = {
        "id": att_id,
        "name": Path(filename).name[:120],
        "ext": ext,
        "kind": kind,
        "bytes": len(data),
        "sha256": hashlib.sha256(data).hexdigest(),
        "created": time.time(),
        "status": "stored",
        "text_chars": 0, "pages": 0, "ocr_pages": 0, "chunks": 0, "small": False,
        "timings_ms": {},
        "guard": None,
    }
    save_meta(meta)
    return meta


def meta_path(att_id: str) -> Path:
    return _dir(att_id) / "meta.json"


def load_meta(att_id: str) -> dict:
    p = meta_path(att_id)
    if not p.exists():
        raise AttachmentError("attachment not found (it may have expired)", status=404)
    return json.loads(p.read_text(encoding="utf-8"))


def save_meta(meta: dict) -> None:
    meta_path(meta["id"]).write_text(json.dumps(meta, indent=2), encoding="utf-8")


def file_path(meta: dict) -> Path:
    return _dir(meta["id"]) / f"file{meta['ext']}"


def text_path(att_id: str) -> Path:
    return _dir(att_id) / "text.txt"


def kb_path(att_id: str) -> Path:
    return _dir(att_id) / "kb.sqlite"


def delete(att_id: str) -> bool:
    d = _dir(att_id)
    if not d.exists():
        return False
    kbstore._cache.pop(str(kb_path(att_id)), None)
    shutil.rmtree(d, ignore_errors=True)
    return True


def sweep_expired(max_age_hours: float = RETENTION_HOURS) -> int:
    """Delete attachments older than the retention window. Returns the count."""
    cutoff = time.time() - max_age_hours * 3600
    removed = 0
    for d in ATTACH_DIR.iterdir():
        if not d.is_dir():
            continue
        try:
            created = json.loads((d / "meta.json").read_text(encoding="utf-8")).get("created", 0)
        except (OSError, ValueError):
            created = d.stat().st_mtime
        if created < cutoff:
            shutil.rmtree(d, ignore_errors=True)
            removed += 1
    return removed


# ------------------------------------------------------------------ OCR

def ocr_available() -> bool:
    return Path(TESSERACT).exists()


def _preprocess_for_ocr(img):
    """Grayscale, upscale small images, stretch contrast. Tesseract's LSTM
    wants ~30 px x-height; phone photos of documents are usually fine, but a
    small screenshot benefits from 2x."""
    from PIL import ImageOps

    img = ImageOps.exif_transpose(img)
    g = img.convert("L")
    w, h = g.size
    if max(w, h) < 1500:
        g = g.resize((w * 2, h * 2))
    g = ImageOps.autocontrast(g, cutoff=1)
    return g


def ocr_image(img, psm: int = 3) -> str:
    """Run Tesseract on a PIL image via a temp PNG. Returns the text ('' if
    Tesseract is missing or fails — OCR is best-effort, never fatal)."""
    if not ocr_available():
        return ""
    import tempfile

    g = _preprocess_for_ocr(img)
    with tempfile.TemporaryDirectory() as td:
        p = Path(td) / "page.png"
        g.save(p, format="PNG")
        try:
            out = subprocess.run(
                [TESSERACT, str(p), "stdout", "--psm", str(psm), "-l", "eng"],
                capture_output=True, text=True, encoding="utf-8", errors="replace",
                timeout=OCR_TIMEOUT_S,
                creationflags=getattr(subprocess, "CREATE_NO_WINDOW", 0))
        except (subprocess.TimeoutExpired, OSError):
            return ""
    if out.returncode != 0:
        return ""
    return _clean_ocr(out.stdout)


_DIGIT_CONFUSIONS = str.maketrans({"@": "0", "O": "0", "o": "0", "l": "1", "I": "1", "|": "1"})
_NUMERIC_TOKEN = re.compile(r"(?<![A-Za-z])(?=[\d@OolI|.,:/\-]*\d)[\d@OolI|][\d@OolI|.,:/\-]*(?![A-Za-z])")


def _fix_digit_confusions(text: str) -> str:
    """Repair the classic OCR swaps (0/O/@, 1/l/I) but ONLY inside tokens that
    are otherwise numeric and not attached to letters, so "3@ days" -> "30 days"
    while "Hello" and "INV-2026" are untouched."""
    def fix(m: re.Match) -> str:
        tok = m.group(0)
        if not any(ch.isdigit() for ch in tok):
            return tok
        return tok.translate(_DIGIT_CONFUSIONS)
    return _NUMERIC_TOKEN.sub(fix, text)


_LIST_START = re.compile(r"^\s*(?:\d+[.)]\s|[-*•]\s)")
_HEADING_LINE = re.compile(r"^[A-Z][A-Z0-9 \-/&'()]{3,}$")


def _unwrap_lines(lines: list[str]) -> list[str]:
    """Re-join sentences that OCR split at the printed line width.

    A scanned policy read "…pattern require" / "30 days written notice…" as two
    lines, and the 1.7B answered "23 days". Joining a line that ends mid-sentence
    (no terminal punctuation) with a continuation that starts in lower case or
    with a digit puts the number back next to its verb. Headings, list items
    and blank lines are never joined."""
    out: list[str] = []
    for ln in lines:
        s = ln.strip()
        prev = out[-1].strip() if out else ""
        if (prev and s and not prev.endswith((".", ":", ";", "!", "?"))
                and (s[0].islower() or s[0].isdigit()) and not _LIST_START.match(s)
                and not _HEADING_LINE.match(prev)):
            out[-1] = out[-1].rstrip() + " " + s
        else:
            out.append(ln)
    return out


def _clean_ocr(text: str) -> str:
    text = text.replace("\f", "\n")
    lines = [ln.rstrip() for ln in text.splitlines()]
    # drop lines that are only OCR noise (1-2 punctuation chars)
    lines = [ln for ln in lines if len(ln.strip()) > 2 or ln.strip().isalnum()]
    lines = _unwrap_lines(lines)
    return _fix_digit_confusions(re.sub(r"\n{3,}", "\n\n", "\n".join(lines)).strip())


# ------------------------------------------------------------------ extraction

def extract(meta: dict) -> dict:
    """Extract text for the stored file, write text.txt, update and return meta."""
    data = file_path(meta).read_bytes()
    t0 = time.time()
    ocr_ms = 0.0
    pages = 0
    ocr_pages = 0
    if meta["kind"] == "text":
        text = chunking.extract_text(meta["name"], data)
    elif meta["kind"] == "pdf":
        text, pages, ocr_pages, ocr_ms = _extract_pdf(data)
    else:  # image
        from PIL import Image

        t = time.time()
        text = ocr_image(Image.open(io.BytesIO(data)))
        ocr_ms = (time.time() - t) * 1000
        pages = 1
        ocr_pages = 1
    text = _normalise(text)
    text_path(meta["id"]).write_text(text, encoding="utf-8")
    meta.update({
        "text_chars": len(text), "pages": pages, "ocr_pages": ocr_pages,
        "ocr_used": ocr_pages > 0, "status": "extracted",
    })
    meta["timings_ms"]["extract"] = round((time.time() - t0) * 1000)
    meta["timings_ms"]["ocr"] = round(ocr_ms)
    save_meta(meta)
    return meta


def _extract_pdf(data: bytes) -> tuple[str, int, int, float]:
    from pypdf import PdfReader

    reader = PdfReader(io.BytesIO(data))
    page_texts: list[str] = []
    for page in reader.pages:
        try:
            page_texts.append(page.extract_text() or "")
        except Exception:  # noqa: BLE001 — a damaged page must not lose the document
            page_texts.append("")
    n = len(page_texts)
    needs_ocr = [i for i, t in enumerate(page_texts) if len(t.strip()) < MIN_TEXT_LAYER_CHARS]
    ocr_ms = 0.0
    ocr_done = 0
    if needs_ocr and ocr_available():
        t = time.time()
        try:
            import pymupdf
            from PIL import Image

            doc = pymupdf.open(stream=data, filetype="pdf")
            for i in needs_ocr[:MAX_OCR_PAGES]:
                pix = doc[i].get_pixmap(dpi=OCR_DPI, colorspace=pymupdf.csGRAY)
                img = Image.frombytes("L", (pix.width, pix.height), pix.samples)
                got = ocr_image(img)
                if got:
                    page_texts[i] = got
                ocr_done += 1
            doc.close()
        except Exception:  # noqa: BLE001 — fall back to whatever text layer exists
            pass
        ocr_ms = (time.time() - t) * 1000
    text = "\n\n".join(f"[page {i + 1}]\n{t}" if n > 1 else t
                       for i, t in enumerate(page_texts) if t.strip())
    return text, n, ocr_done, ocr_ms


def _normalise(text: str) -> str:
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    text = re.sub(r"[ \t]+\n", "\n", text)
    return text.strip()


# ------------------------------------------------------------------ indexing

def index(meta: dict, embedder_id: str | None = None) -> dict:
    """Chunk + embed the extracted text into the attachment's own kb.sqlite."""
    text = text_path(meta["id"]).read_text(encoding="utf-8") if text_path(meta["id"]).exists() else ""
    t0 = time.time()
    chunks = chunking.split_text(text) if text else []
    meta["small"] = len(text) <= FULLTEXT_CHARS
    if chunks and not meta["small"]:
        # Same embedder as the agent's own KB, so one query vector serves both.
        vecs = embeddings.embed_passages(chunks, embedder_id)
        kb = kbstore.connect(kb_path(meta["id"]))
        kbstore.add_document(kb, meta["name"], chunks, vecs, meta={"bytes": meta["bytes"]},
                             embedder_id=embedder_id)
        kb.close()
    meta["chunks"] = len(chunks)
    meta["embedder"] = embedder_id
    meta["status"] = "ready" if text else "no-text"
    meta["timings_ms"]["index"] = round((time.time() - t0) * 1000)
    save_meta(meta)
    return meta


def full_text(att_id: str) -> str:
    p = text_path(att_id)
    return p.read_text(encoding="utf-8") if p.exists() else ""


def context_for(meta: dict, query: str, top_k: int = ATTACH_TOP_K,
                min_score: float = ATTACH_MIN_SCORE) -> list[dict]:
    """On-demand retrieval over one attachment.

    Small document -> every chunk, no embedding (the whole file fits).
    Large document -> lexical gate, then vector search with a low floor.
    Result rows look like kbstore.search rows plus kind='attachment'.
    """
    text = full_text(meta["id"])
    if not text:
        return []
    if meta.get("small"):
        chunks = chunking.split_text(text) or [text]
        return [{"chunk_id": -1, "doc_name": meta["name"], "chunk_index": i, "text": c,
                 "score": 1.0, "kind": "attachment"} for i, c in enumerate(chunks)]
    ok, _ = gating.should_retrieve(query)
    if not ok:
        return []
    kbp = kb_path(meta["id"])
    if not kbp.exists():
        return []
    kb = kbstore.connect(kbp)
    try:
        hits = kbstore.search(kb, embeddings.embed_query(query, meta.get("embedder")), top_k=top_k, min_score=min_score)
    finally:
        kb.close()
    for h in hits:
        h["kind"] = "attachment"
    return hits


def public(meta: dict) -> dict:
    """The attachment as the UI sees it — never the raw text or findings' values."""
    out = {k: meta.get(k) for k in (
        "id", "name", "ext", "kind", "bytes", "status", "text_chars", "pages",
        "ocr_pages", "ocr_used", "chunks", "small", "timings_ms", "created")}
    g = meta.get("guard")
    out["guard"] = g if g is None else {k: g.get(k) for k in (
        "enabled", "blocked", "summary", "judged", "judge_note", "ms", "message")}
    return out

File: android_app/AgentRuntime/src/attachments.ts
// Per-conversation file attachments on the handset — the on-device port of
// runtime/attachments.py.
//
// One file per turn: jpg/jpeg/pdf/txt/md, 5 MB max, checked by extension AND
// by the file's first bytes. Text is read directly; images are OCR'd with
// ML Kit (on-device, no network); PDFs are rendered page by page with the
// platform's PdfRenderer and OCR'd the same way — one uniform path for text
// PDFs and scans alike, capped at MAX_PDF_PAGES to bound the wall-clock.
//
// "On-demand RAG": a small document (<= FULLTEXT_CHARS) is injected whole at
// question time; a larger one is chunked with the same section-aware splitter
// as the KB, embedded with the same bge-small model, and searched per question
// behind the same lexical gate. Attachment chunks live in their own SQLite
// file beside the copied file, never in the agent's inline KB — an attachment
// is conversation-scoped user data, not knowledge, and is swept after
// RETENTION_HOURS.
//
// The PII Guard verdict is stored on the attachment so every later question
// against a blocked file is refused without rescanning (see chat.ts).
import { open, DB } from '@op-engineering/op-sqlite';
import ReactNativeBlobUtil from 'react-native-blob-util';
import { pick, types } from '@react-native-documents/picker';
import TextRecognition from '@react-native-ml-kit/text-recognition';
import PdfThumbnail from 'react-native-pdf-thumbnail';

import { AGENTS_DIR } from './config';
import { DEFAULT_EMBEDDER, EmbedderSpec } from './embedder';
import { splitText } from './chunker';
import { shouldRetrieve } from './gating';
import { embedText } from './llm';
import { SearchHit } from './kb';
import { Verdict } from './guard';
import { log } from './logger';

const fs = ReactNativeBlobUtil.fs;

export const ALLOWED_EXT: Record<string, 'image' | 'pdf' | 'text'> = {
  '.jpg': 'image', '.jpeg': 'image', '.pdf': 'pdf', '.txt': 'text', '.md': 'text',
};
export const MAX_ATTACH_BYTES = 5 * 1024 * 1024;
export const RETENTION_HOURS = 24;
export const FULLTEXT_CHARS = 3500;
export const ATTACH_TOP_K = 4;
export const ATTACH_MIN_SCORE = 0.30;
export const MAX_PDF_PAGES = 20;
const PDF_RENDER_QUALITY = 85;

export type PickedFile = { name: string; size: number; uri: string };

export type Attachment = {
  id: string;
  agentId: string;
  name: string;
  ext: string;
  kind: 'image' | 'pdf' | 'text';
  bytes: number;
  dir: string;
  path: string;              // copied file, always a plain filesystem path
  text: string;              // extracted text (kept in memory for the session)
  textChars: number;
  pages: number;
  ocrPages: number;
  chunks: number;
  small: boolean;
  status: 'ready' | 'no-text';
  guardEnabled: boolean;
  guard: Verdict | null;
  judgedChunks: number[];
  timingsMs: Record<string, number>;
  embedder: EmbedderSpec;      // the agent's embedder, so KB and file share one query vector
};

export class AttachmentError extends Error {}

// ------------------------------------------------------------------ picking

/** Open the system picker limited to the supported types. null = cancelled. */
export async function pickAttachment(): Promise<PickedFile | null> {
  try {
    const [res] = await pick({
      mode: 'open',
      allowMultiSelection: false,
      type: [types.pdf, types.images, types.plainText, 'text/markdown', 'text/x-markdown'],
    });
    if (!res) return null;
    return { name: res.name || 'attachment', size: Number(res.size || 0), uri: res.uri };
  } catch (e: any) {
    if (/cancel/i.test(String(e?.message || e))) return null;
    throw e;
  }
}

export function extOf(name: string): string {
  const m = /\.[A-Za-z0-9]+$/.exec(name || '');
  return m ? m[0].toLowerCase() : '';
}

/** Extension + size check — the cheap half of validation, before any copy. */
export function validateMeta(file: PickedFile): { ext: string; kind: 'image' | 'pdf' | 'text' } {
  const ext = extOf(file.name);
  const kind = ALLOWED_EXT[ext];
  if (!kind) throw new AttachmentError(`'${ext || file.name}' is not supported — attach a jpg, jpeg, pdf, txt or md file`);
  if (file.size <= 0) throw new AttachmentError('the file is empty');
  if (file.size > MAX_ATTACH_BYTES) {
    throw new AttachmentError(`file is ${(file.size / 1048576).toFixed(1)} MB; the limit is 5 MB`);
  }
  return { ext, kind };
}

const B64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';

/** Decode the first `max` bytes of a base64 string (React Native has no Buffer). */
export function base64Head(b64: string, max: number): number[] {
  const out: number[] = [];
  let bits = 0;
  let acc = 0;
  for (let i = 0; i < b64.length && out.length < max; i++) {
    const v = B64.indexOf(b64[i]);
    if (v < 0) continue;          // '=' padding or whitespace
    acc = (acc << 6) | v;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out.push((acc >> bits) & 0xff);
    }
  }
  return out;
}

/** Magic-byte check on the copied file — a renamed binary is refused, not guessed at. */
async function validateBytes(path: string, kind: 'image' | 'pdf' | 'text'): Promise<void> {
  const b64 = await fs.readFile(path, 'base64');
  const head = base64Head(String(b64).slice(0, 8192), 4096);
  if (kind === 'image' && !(head[0] === 0xff && head[1] === 0xd8 && head[2] === 0xff)) {
    throw new AttachmentError('this is not a JPEG image (the bytes do not match the extension)');
  }
  if (kind === 'pdf') {
    const ascii = String.fromCharCode(...head.slice(0, 1024)).trimStart();
    if (!ascii.startsWith('%PDF-')) {
      throw new AttachmentError('this is not a PDF (the bytes do not match the extension)');
    }
  }
  if (kind === 'text' && head.includes(0)) {
    throw new AttachmentError('this is not a text file (binary content)');
  }
}

// ------------------------------------------------------------------ storage

const attachRoot = (agentId: string) => `${AGENTS_DIR}/${agentId}/attachments`;

function newId(): string {
  return Date.now().toString(36) + Math.random().toString(36).slice(2, 10);
}

async function copyIn(src: string, dest: string): Promise<void> {
  // content:// and file:// both read through RNBU; a 5 MB base64 round-trip is
  // the simplest path that works for every provider the picker can return.
  const from = src.startsWith('content://') ? src : src.replace('file://', '');
  const b64 = await fs.readFile(from, 'base64');
  await fs.writeFile(dest, b64, 'base64');
}

export async function removeAttachment(att: Attachment): Promise<void> {
  await fs.unlink(att.dir).catch(() => {});
}

/** Delete attachments older than the retention window for this agent. */
export async function sweepAttachments(agentId: string): Promise<number> {
  const root = attachRoot(agentId);
  if (!(await fs.exists(root))) return 0;
  let removed = 0;
  const cutoff = Date.now() - RETENTION_HOURS * 3600 * 1000;
  for (const id of await fs.ls(root)) {
    const dir = `${root}/${id}`;
    try {
      const st: any = await fs.stat(dir);
      if (Number(st.lastModified) < cutoff) { await fs.unlink(dir); removed++; }
    } catch {}
  }
  if (removed) await log(`attach: swept ${removed} expired attachment(s)`);
  return removed;
}

// ------------------------------------------------------------------ OCR text repair
// Ports of runtime/attachments._fix_digit_confusions and _unwrap_lines.

const DIGIT_CONFUSIONS: Record<string, string> = { '@': '0', O: '0', o: '0', l: '1', I: '1', '|': '1' };
const NUMERIC_TOKEN = /(?<![A-Za-z])(?=[\d@OolI|.,:/\-]*\d)[\d@OolI|][\d@OolI|.,:/\-]*(?![A-Za-z])/g;

export function fixDigitConfusions(text: string): string {
  return text.replace(NUMERIC_TOKEN, tok =>
    /\d/.test(tok) ? tok.replace(/[@OolI|]/g, c => DIGIT_CONFUSIONS[c]) : tok);
}

const LIST_START = /^\s*(?:\d+[.)]\s|[-*•]\s)/;
const HEADING_LINE = /^[A-Z][A-Z0-9 \-/&'()]{3,}$/;

export function unwrapLines(lines: string[]): string[] {
  const out: string[] = [];
  for (const ln of lines) {
    const s = ln.trim();
    const prev = out.length ? out[out.length - 1].trim() : '';
    if (prev && s && !/[.:;!?]$/.test(prev) && /^[a-z0-9]/.test(s)
        && !LIST_START.test(s) && !HEADING_LINE.test(prev)) {
      out[out.length - 1] = out[out.length - 1].trimEnd() + ' ' + s;
    } else {
      out.push(ln);
    }
  }
  return out;
}

export function cleanOcr(text: string): string {
  let lines = text.replace(/\f/g, '\n').split(/\r?\n/).map(l => l.trimEnd());
  lines = lines.filter(l => l.trim().length > 2 || /^[A-Za-z0-9]+$/.test(l.trim()));
  lines = unwrapLines(lines);
  return fixDigitConfusions(lines.join('\n').replace(/\n{3,}/g, '\n\n')).trim();
}

// ------------------------------------------------------------------ extraction

async function ocrImage(path: string): Promise<string> {
  const res = await TextRecognition.recognize(path.startsWith('file://') ? path : `file://${path}`);
  // Blocks preserve reading order better than the flat `text` on multi-column pages.
  const lines: string[] = [];
  for (const block of res.blocks || []) {
    for (const line of block.lines || []) lines.push(line.text);
    lines.push('');
  }
  return cleanOcr(lines.length ? lines.join('\n') : (res.text || ''));
}

async function extractPdf(path: string): Promise<{ text: string; pages: number }> {
  const parts: string[] = [];
  let pages = 0;
  for (let i = 0; i < MAX_PDF_PAGES; i++) {
    let thumb;
    try {
      thumb = await PdfThumbnail.generate(`file://${path}`, i, PDF_RENDER_QUALITY);
    } catch {
      break;   // past the last page (the renderer throws for an out-of-range index)
    }
    pages++;
    try {
      const t = await ocrImage(thumb.uri);
      if (t.trim()) parts.push(`[page ${i + 1}]\n${t}`);
    } finally {
      await fs.unlink(thumb.uri.replace('file://', '')).catch(() => {});
    }
  }
  return { text: parts.join('\n\n'), pages };
}

// ------------------------------------------------------------------ indexing

const SCHEMA = `CREATE TABLE IF NOT EXISTS chunks (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  chunk_index INTEGER NOT NULL,
  text TEXT NOT NULL,
  embedding BLOB NOT NULL
);`;

function openDb(att: Attachment): DB {
  return open({ name: 'attach.sqlite', location: att.dir });
}

async function indexChunks(att: Attachment, chunks: string[], onStatus?: (s: string) => void): Promise<void> {
  const spec = att.embedder;
  const db = openDb(att);
  try {
    await db.execute(SCHEMA);
    for (let i = 0; i < chunks.length; i++) {
      onStatus?.(`Indexing ${att.name} — chunk ${i + 1}/${chunks.length}…`);
      const vec = await embedText(chunks[i], false, spec);
      await db.execute('INSERT INTO chunks (chunk_index, text, embedding) VALUES (?, ?, ?)',
        [i, chunks[i], vec.buffer as ArrayBuffer]);
    }
  } finally {
    db.close();
  }
}

function blobToVec(blob: any): Float32Array | null {
  if (blob instanceof ArrayBuffer) return new Float32Array(blob);
  if (blob?.buffer instanceof ArrayBuffer) {
    const bytes = Number(blob.byteLength ?? blob.buffer.byteLength - (blob.byteOffset ?? 0));
    return new Float32Array(blob.buffer, blob.byteOffset ?? 0, Math.floor(bytes / 4));
  }
  return null;
}

async function searchAttachment(att: Attachment, query: Float32Array, topK: number, minScore: number): Promise<SearchHit[]> {
  const db = openDb(att);
  try {
    const res = await db.execute('SELECT chunk_index, text, embedding FROM chunks ORDER BY chunk_index');
    const scored: Array<{ idx: number; text: string; score: number }> = [];
    for (const row of res.rows ?? []) {
      const vec = blobToVec((row as any).embedding);
      if (!vec || vec.length !== query.length) continue;
      let dot = 0;
      for (let i = 0; i < query.length; i++) dot += vec[i] * query[i];
      if (dot >= minScore) scored.push({ idx: Number((row as any).chunk_index), text: String((row as any).text), score: dot });
    }
    scored.sort((a, b) => b.score - a.score);
    return scored.slice(0, topK).map(s => ({
      doc_name: att.name, chunk_index: s.idx, text: s.text,
      score: Math.round(s.score * 1e4) / 1e4, source: 'attachment' as const,
    }));
  } finally {
    db.close();
  }
}

/** On-demand retrieval over one attachment (see runtime/attachments.context_for). */
export async function contextFor(att: Attachment, query: string): Promise<SearchHit[]> {
  if (!att.text) return [];
  if (att.small) {
    const chunks = splitText(att.text);
    return (chunks.length ? chunks : [att.text]).map((c, i) => ({
      doc_name: att.name, chunk_index: i, text: c, score: 1.0, source: 'attachment' as const,
    }));
  }
  const [ok] = shouldRetrieve(query);
  if (!ok) return [];
  const qvec = await embedText(query, true, att.embedder);
  return searchAttachment(att, qvec, ATTACH_TOP_K, ATTACH_MIN_SCORE);
}

// ------------------------------------------------------------------ ingest

/** validate -> copy -> extract (OCR) -> chunk (+embed when large). The guard
 *  runs afterwards in chat.ts, where the agent's model is available. */
export async function ingest(agentId: string, file: PickedFile,
                             onStatus?: (s: string) => void,
                             embedder: EmbedderSpec = DEFAULT_EMBEDDER): Promise<Attachment> {
  const { ext, kind } = validateMeta(file);
  const id = newId();
  const dir = `${attachRoot(agentId)}/${id}`;
  await fs.mkdir(attachRoot(agentId)).catch(() => {});
  await fs.mkdir(dir).catch(() => {});
  const path = `${dir}/file${ext}`;
  const timingsMs: Record<string, number> = {};

  try {
    onStatus?.(`Copying ${file.name}…`);
    let t = Date.now();
    await copyIn(file.uri, path);
    await validateBytes(path, kind);
    timingsMs.copy = Date.now() - t;

    t = Date.now();
    let text = '';
    let pages = 0;
    let ocrPages = 0;
    if (kind === 'text') {
      text = String(await fs.readFile(path, 'utf8'));
    } else if (kind === 'image') {
      onStatus?.('Reading the image (OCR)…');
      text = await ocrImage(path);
      pages = 1; ocrPages = 1;
    } else {
      onStatus?.('Rendering and reading PDF pages (OCR)…');
      const r = await extractPdf(path);
      text = r.text; pages = r.pages; ocrPages = r.pages;
    }
    text = text.replace(/\r\n?/g, '\n').replace(/[ \t]+\n/g, '\n').trim();
    timingsMs.extract = Date.now() - t;

    const att: Attachment = {
      id, agentId, name: file.name.slice(0, 120), ext, kind, bytes: file.size, dir, path,
      text, textChars: text.length, pages, ocrPages, chunks: 0,
      small: text.length <= FULLTEXT_CHARS, status: text ? 'ready' : 'no-text',
      guardEnabled: false, guard: null, judgedChunks: [], timingsMs, embedder,
    };

    t = Date.now();
    const chunks = text ? splitText(text) : [];
    att.chunks = chunks.length;
    if (chunks.length && !att.small) await indexChunks(att, chunks, onStatus);
    timingsMs.index = Date.now() - t;

    await log(`attach: ${att.name} ${kind} ${att.bytes}B -> ${att.textChars} chars, ${pages} pages, ` +
              `ocr=${ocrPages}, ${att.chunks} chunks, small=${att.small} ` +
              `(copy ${timingsMs.copy}ms, extract ${timingsMs.extract}ms, index ${timingsMs.index}ms)`);
    return att;
  } catch (e) {
    await fs.unlink(dir).catch(() => {});
    throw e;
  }
}
